File: portal/frontend/src/services/youtube.ts
import type { YouTubeStatus, StreamStats, UpdateStreamRequest } from '@/types/youtube'

const API_BASE = 'http://localhost:8080'

export const youtubeAPI = {
  async getStatus(): Promise<YouTubeStatus> {
    const res = await fetch(`${API_BASE}/api/youtube/status`)
    if (!res.ok) throw new Error('Failed to fetch YouTube status')
    return res.json()
  },

  async getStats(): Promise<StreamStats> {
    const res = await fetch(`${API_BASE}/api/youtube/stats`)
    if (!res.ok) throw new Error('Failed to fetch YouTube stats')
    return res.json()
  },

  async updateStream(data: UpdateStreamRequest): Promise<void> {
    const res = await fetch(`${API_BASE}/api/youtube/stream`, {
      method: 'PATCH',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(data),
    })
    if (!res.ok) throw new Error('Failed to update stream')
  },
}

File: portal/backend/internal/api/handlers/youtube.go
package handlers

import (
	"encoding/json"
	"net/http"

	"kit.workmate/live-portal/internal/services/youtube"
)

type YouTubeHandler struct {
	client *youtube.Client
}

func NewYouTubeHandler(client *youtube.Client) *YouTubeHandler {
	return &YouTubeHandler{
		client: client,
	}
}

// GetStatus returns YouTube connection status
func (h *YouTubeHandler) GetStatus(w http.ResponseWriter, r *http.Request) {
	if h.client == nil {
		http.Error(w, "YouTube not configured", http.StatusServiceUnavailable)
		return
	}

	status := h.client.GetStatus()

	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(status)
}

// GetStats returns YouTube stream statistics
func (h *YouTubeHandler) GetStats(w http.ResponseWriter, r *http.Request) {
	if h.client == nil {
		http.Error(w, "YouTube not configured", http.StatusServiceUnavailable)
		return
	}

	stats, err := h.client.GetStats()
	if err != nil {
		http.Error(w, err.Error(), http.StatusInternalServerError)
		return
	}

	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(stats)
}

// UpdateStream updates stream metadata (title, description)
func (h *YouTubeHandler) UpdateStream(w http.ResponseWriter, r *http.Request) {
	if h.client == nil {
		http.Error(w, "YouTube not configured", http.StatusServiceUnavailable)
		return
	}

	var req youtube.UpdateStreamRequest
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		http.Error(w, "Invalid request body", http.StatusBadRequest)
		return
	}

	if err := h.client.UpdateStream(&req); err != nil {
		http.Error(w, err.Error(), http.StatusInternalServerError)
		return
	}

	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(map[string]string{
		"status": "success",
	})
}
